Refactor the AD drivers classes

jppelteret opened the pull request:
This PR significantly reworks the implementation of the AD driver classes. Firstly, all class methods are moved to the associated source file. The (logically) static class methods have been removed, so all methods must be called through an instance of a driver class. The previously implemented logic for existent ADOL-C tapes has been superseded by low-level calls to the library itself.

The ADHelper classes are updated to accommodate the changes to the driver classes. Further features provided by the drivers are exposed through the base helper class.

Fixes #7335
Fixes #7372

jppelteret commented:
Most of the changes in the PR result from migrating class methods from `ad_drivers.h` to `ad_drivers.cc`.

Sorry for the noise generated by having yet another PR on this topic. I sincerely hope that this is the last time that I touch these classes for the next long while.

jppelteret commented:
/run-tests

FYI. With these changes all `adolc` and `sacado` tests still pass for me.

jppelteret commented:
Thanks for the fixes @masterleinad!

jppelteret commented on source/differentiation/ad/ad_drivers.cc:
```diff
@@ -0,0 +1,2260 @@
+        {
+          return false;
+        }
+    }
```
This is the fix for #7372 (i.e. reverting this function definition to its state before the last change.)

jppelteret commented on source/differentiation/ad/ad_drivers.cc:
```diff
@@ -0,0 +1,2260 @@
+      Assert(status_tape < 4 && status_tape >= -2,
+             ExcIndexRange(status_tape, -2, 4));
+      return (status_tape < 0);
+    }
```
New feature.

jppelteret commented on source/differentiation/ad/ad_drivers.cc:
```diff
@@ -0,0 +1,2260 @@
+      last_action_requires_retaping() const
+    {
+      return requires_retaping(active_tape);
+    }
```
New feature.

jppelteret commented on source/differentiation/ad/ad_drivers.cc:
```diff
@@ -0,0 +1,2260 @@
+             ExcMessage(
+               "This tape does not exist, and therefore cannot be cleared."));
+      removeTape(tape_index, TapeRemovalType::ADOLC_REMOVE_COMPLETELY);
+    }
```
New feature.

jppelteret commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -275,6 +275,39 @@ namespace Differentiation
+       */
+      bool
+      requires_retaping(
+        const typename Types<ADNumberType>::tape_index tape_index) const;
```
New feature -> Documentation requires review.

jppelteret commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -275,6 +275,39 @@ namespace Differentiation
+       * evaluation point.
+       */
+      bool
+      last_action_requires_retaping() const;
```
New feature -> Documentation requires review.

bangerth replied:
The documentation of this function appears identical to me to the one of the previous function. How do they differ?

jppelteret commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -258,6 +252,12 @@ namespace Differentiation
+       * Return a list of registered tape indices.
+       */
+      std::vector<typename Types<ADNumberType>::tape_index>
+      get_registered_tape_indices() const;
```
New feature -> Documentation requires review.

jppelteret commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -275,6 +275,39 @@ namespace Differentiation
+       * Completely erases the tape with the given @p tape_index.
+       */
+      void
+      remove_tape(const typename Types<ADNumberType>::tape_index tape_index);
```
New feature -> Documentation requires review.

jppelteret commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -1171,6 +845,33 @@ namespace Differentiation
+       * When the @p status variable takes a negative value, retaping of the dependent
+       * function is necessary for a reliable computation to be performed.
+       */
+      mutable std::map<typename Types<ADNumberType>::tape_index, int> status;
```
New attribute -> Documentation requires review.

jppelteret commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -475,6 +478,50 @@ namespace Differentiation
+       */
+      bool
+      recorded_tape_requires_retaping(
+        const typename Types<ad_type>::tape_index tape_index) const;
```
New feature -> Documentation requires review.

jppelteret commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -475,6 +478,50 @@ namespace Differentiation
+       * or derivatives of the dependent variables).
+       */
+      bool
+      active_tape_requires_retaping() const;
```
New feature -> Documentation requires review.

jppelteret commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -475,6 +478,50 @@ namespace Differentiation
+       * the active_tape_requires_retaping() function.
+       */
+      void
+      clear_active_tape();
```
New feature -> Documentation requires review.

jppelteret commented on source/differentiation/ad/ad_drivers.cc:
```diff
@@ -0,0 +1,2260 @@
+                registered_tape_indices_s.end(),
+                registered_tape_indices.begin());
+      return registered_tape_indices;
+    }
```
We now get this information from ADOL-C directly, rather than keeping track of it ourselves.

bangerth replied:
nice!

jppelteret commented on source/differentiation/ad/ad_drivers.cc:
```diff
@@ -0,0 +1,2260 @@
+          stream << tape_index << "->" << status_tape
+                 << (i < (registered_tape_indices.size() - 1) ? "," : "");
+        }
+      stream << "\n";
```
This listing of registered tapes was reworked to accommodate the revised data structures of this class.

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -475,6 +478,50 @@ namespace Differentiation
+       * evaluation point.
+       *
+       * For the output of this function to be meaningful, it must be called
+       * after the activate_recorded_tape() is called and the new evaluation
```
```suggestion
       * after activate_recorded_tape() is called and the new evaluation
```

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -475,6 +478,50 @@ namespace Differentiation
+       * computation to be performed on a tape with the given @p tape_index.
+       * This may be necessary a sign comparison within branched operations
+       * yields different results to those computed at the original tape
+       * evaluation point.
```
same comments about this sentence as before

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -475,6 +478,50 @@ namespace Differentiation
+       * retaping of the dependent function is necessary for a reliable
+       * computation to be performed on the avtive tape. This may be necessary a
+       * sign comparison within branched operations yields different results to
+       * those computed at the original tape evaluation point.
```
and here

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -475,6 +478,50 @@ namespace Differentiation
+       * those computed at the original tape evaluation point.
+       *
+       * For the output of this function to be meaningful, it must be called
+       * after the activate_recorded_tape() is called and the new evaluation
```
```suggestion
       * after activate_recorded_tape() is called and the new evaluation
```

bangerth commented on source/differentiation/ad/ad_drivers.cc:
```diff
@@ -0,0 +1,2260 @@
+        registered_tape_indices(registered_tape_indices_s.size());
+      std::copy(registered_tape_indices_s.begin(),
+                registered_tape_indices_s.end(),
+                registered_tape_indices.begin());
```
You can do this more easily by constructing straight from an iterator range:
```
  return std::vector<...> (registered_tape_indices_s.begin(), registered_tape_indices_s.end());
```
See constructor (4) here: https://en.cppreference.com/w/cpp/container/vector/vector

jppelteret replied:
Thanks for the tip!

bangerth commented on source/differentiation/ad/ad_drivers.cc:
```diff
@@ -0,0 +1,2260 @@
+      // See ADOL-C manual section 1.7 and comments in last paragraph of
+      // section 3.1
+      Assert(status_tape < 4 && status_tape >= -2,
+             ExcIndexRange(status_tape, -2, 4));
```
I would rather we use `ExcInternalError()` or a custom message for this. That's because we use the exception you use here for user-provided indices, but here if you get something out of range, there is nothing a user could have done wrong...

bangerth commented on source/differentiation/ad/ad_drivers.cc:
```diff
@@ -0,0 +1,2260 @@
+                 double,
+                 typename std::enable_if<
+                   ADNumberTraits<ADNumberType>::type_code ==
+                   NumberTypes::adolc_taped>::type>::print(std::ostream &stream)
```
sigh :-( This indentation is entirely unreadable...

jppelteret replied:
Yes, this is slightly unfortunate :-(

bangerth commented:
Let me know if you have news!

jppelteret commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -1171,6 +845,33 @@ namespace Differentiation
+       * - <b>-1:</b> An <code>adouble</code> comparison yields different
+       *     results from the evaluation point at which the tape was generated.
+       * - <b>-2:</b> The argument of a user-defined quadrature has changed from
+       *     the evaluation point at which the tape was generated.
```
These are literally quadrature rules for some special integration functions that ADOL-C supports. Apparently...

jppelteret commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -275,6 +275,39 @@ namespace Differentiation
+       * a reliable computation to be performed.
+       * This may be necessary a sign comparison within branched operations
+       * yields different results to those computed at the original tape
+       * evaluation point.
```
Ok, I've added some more detail and a trivial example for this. Its duplicated at all the points you've noted this issue.

jppelteret commented:
@bangerth I know this took me forever to get to, so I really apologise about that. I think that I've addressed all of your comments. Thanks very much for your input!

jppelteret commented:
/rebuild

jppelteret commented:
@bangerth ping :-)

bangerth commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -275,6 +275,81 @@ namespace Differentiation
+       *     return y;
+       *   else
+       *     return x*z;
+       * }
```
Thanks for adding this example! 

In order for the system to determine that retaping is necessary, I guess it doesn't just record the parse tree of the operations applied, but also that at one point it checked the condition `x<y`? Without remembering this condition, it would not be able to tell that retaping is necessary.

If I'm correct with this, could you add a sentence to this effect to the end of the paragraph following this example?

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -475,6 +478,93 @@ namespace Differentiation
+       *   else
+       *     return x*z;
+       * }
+       * @endcode
```
same issue here (and about the next function below) as commented on above

jppelteret commented:
/rebuild

jppelteret commented:
@bangerth Thanks again for the review. I think that I managed to address the last of your concerns.

bangerth commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -275,6 +275,87 @@ namespace Differentiation
+       * applied on the branch chosen at the time to taping, but also checks
+       * that the condition continues to be satisfied. For some other evaluation
+       * of the tape (i.e. for some different
+       * inputs @p x and @p y, the other branch of the conditional check may be
```
closing paren missing here

bangerth commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -275,6 +275,87 @@ namespace Differentiation
+       * chosen. The result of following this code path has not been recorded on
+       * the tape, and therefore cannot be evaluated. In such a case, it is
+       * therefore necessary to re-record the tape at the new evaluation point
+       * in order to resolve the new code branch.
```
maybe say "In such a case, the underlying AD library will be able to tell you that it is necessary to re-record the tape ...branch. This function can be used to find out whether this is so."

Good case why using passive statements ("it is necessary") should be avoided :-)

bangerth replied:
same comments for the function immediately above

jppelteret commented:
/rebuild
